engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java

```diff
@@ -0,0 +1,156 @@
+package org.enso.interpreter.node.expression.builtin;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Bind;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.data.vector.ArrayLikeHelpers;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * Base class for every Enso builtin object. Not type. Note that base class for a builtin type is
+ * {@link Builtin}.
+ *
+ * <p>The {@link InteropLibrary interop} protocol roughly corresponds to the implementation of the
+ * protocol inside {@link org.enso.interpreter.runtime.data.atom.Atom}.
+ *
+ * <p>Note that extension methods are not resolved, because they are not defined in builtins module
+ * scope. In other words, extension methods are not reported as members via interop.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+public abstract class BuiltinObject extends EnsoObject {
+
+  private final String builtinName;
+  private Map<String, Function> methods;
+  private Builtin cachedBuiltinType;
+
+  /**
+   * @param builtinName Simple name of the builtin that should be contained in {@link
+   *     org.enso.interpreter.runtime.builtin.Builtins#builtinsByName}.
```

Comment:
Out of scope for this PR, as it would require changes to how [Builtins](https://github.com/enso-org/enso/blob/d687365070381b07eb4214bc665a13912c3bbb96/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java) is initialized. But I share your gut feeling.